nx/test/nx/defn/grad_test.exs

```diff
@@ -2503,6 +2503,101 @@ defmodule Nx.Defn.GradTest do
     end
   end
 
+  describe "take_along_axis" do
+    defn grad_sum_take_along_axis(t, i) do
+      grad(
+        t,
+        fn t ->
+          t
+          |> Nx.take_along_axis(i, axis: 1)
+          |> Nx.sum()
+        end
+      )
+    end
+
+    defn grad_sum_take_along_axis_power(t, i) do
+      grad(
+        t,
+        fn t ->
+          t
+          |> Nx.power(2)
+          |> Nx.take_along_axis(i, axis: 1)
+          |> Nx.sum()
+        end
+      )
+    end
+
+    defn grad_sum_log_power_take_along_axis_cos(t, i) do
+      grad(
+        t,
+        fn t ->
+          t
+          |> Nx.cos()
+          |> Nx.take_along_axis(i, axis: 1)
+          |> Nx.power(2)
+          |> Nx.log()
+          |> Nx.sum()
+        end
+      )
+    end
+
+    test "computes gradient" do
+      assert Nx.tensor([
+               [3.0, 0.0],
+               [1.0, 2.0],
+               [1.0, 2.0]
+             ]) ==
+               grad_sum_take_along_axis(
+                 Nx.tensor([
+                   [0, 1],
+                   [2, 3],
+                   [4, 5]
+                 ]),
+                 Nx.tensor([
+                   [0, 0, 0],
+                   [1, 1, 0],
+                   [0, 1, 1]
+                 ])
+               )
+
+      assert Nx.tensor([
+               [0.0, 0.0],
+               [4.0, 12.0],
+               [8.0, 20.0]
+             ]) ==
+               grad_sum_take_along_axis_power(
+                 Nx.tensor([
+                   [0, 1],
+                   [2, 3],
+                   [4, 5]
+                 ]),
+                 Nx.tensor([
+                   [0, 0, 0],
+                   [1, 1, 0],
+                   [0, 1, 1]
+                 ])
+               )
+
+      assert Nx.tensor([
+               [-0.0, -0.0],
+               [4.370079, 0.5701862],
```

Comment:
@josevalim This assertion is failing in a rather bizarre way. The actual tensor returned is:

```elixir
#Nx.Tensor<
  f32[3][2]
  [
    [-0.0, -0.0],
    [1.8369784355163574, 0.570186197757721],
    [-2.3156425952911377, 13.522059440612793]
  ]
>
```

which only differs from the current tensor on the index `[1, 0]`, with actual value `1.8` instead of `4.3`...

Any ideas? Also, this might or might not be related to me needing to call `Nx.gather` to actually propagate the chain rule from the outermost functions, although it seemed to make sense.

Reply:
The other two assertions are working as intended :)

Reply:
Why do we do gather actually? Shouldn't we use `Nx.reshape(g, {num_elements})` as the updates? I.e. we take every gradient element and put it into wherever element it originated from.

Reply:
Huh, this could work! Gather was just more instinctive, but they might be equivalent here :)

Reply:
Well, actually the scatter indices are valid indices into the original tensor, not into the grad tensor. To give an example, the original tensor may very well shrink after `take_along_axis`, in which case the grad tensor is smaller than the original tensor and the indexing is different :)

Reply:
Right, so maybe this is what's causing the strange error I'm getting

Reply:
Just for curiosity, those values came from JAX, right?

Reply:
@josevalim yes

Reply:
As commented above, reshaping `g` to a `{num_elements}`-shaped tensor did it